# LOADING: Packages and function definitions
```{r}
library(data.table)
library(ggplot2)
source("~/Documents/projects/pr_integration/wd_integration/Code/UtillityScripts/PlottingFunctions.r")

rotate <- function(x) t(apply(x, 2, rev))

custom.gls <- function(x, y, Sigi) {
  x <- as.matrix(x, ncol=1)
  y <- as.matrix(y, ncol=1)

  # Calculated beta on pre-computed inverse matrix
  # xtxi = b1
  xtxi <- solve(t(x) %*% Sigi %*% x)
  #  part between brackets is b2
  beta <- xtxi %*% (t(x) %*% Sigi %*% y)
  
  # Calculate SE
  res <- y - (x %*% beta)
  sig <- sqrt(sum(res^2) / (nrow(x)-2))
  se  <- sqrt(diag(xtxi))*sig
  
  # Calculate p
  tstats <- abs(beta / se)
  pval <- 2 * pt(tstats, df=nrow(x)-2, lower=F)
  return(c(beta, se, pval))
}

custom.gls.intercept <- function(x, y, Sigi) {
  # Add intercept
  x <- cbind(rep(1, length(x)), x)

  # Calculated beta on pre-computed inverse matrix
  xtxi <- solve(t(x) %*% Sigi %*% x )
  beta <- xtxi %*% t(x) %*% Sigi %*% y
  
  # Calculate SE
  res <- y - (x %*% beta)
  sig <- sqrt(sum(res^2) / (nrow(x)-2))
  se  <- sqrt(diag(xtxi))*sig
  
  # Calculate p
  tstats <- abs(beta / se)
  pval <- 2 * pt(tstats[2], df=nrow(x)-2, lower=F)
  
  return(c(beta[2], se[2], pval))
}

custom.gls.per.arm <- function(x, y, Sigi) {
  b1 <- t(x) %*% Sigi %*% x
  b2 <- t(x) %*% Sigi %*% y
  return(c(b1, b2))
}
```

# LOADING: Data loading
```{r}
# Ruwe GWAS gene pvalues
gwas.x <- data.frame(fread("~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/height_2018_30124842_hg19_genePvalues.txt", data.table = F), row.names = 1)

gwas.x.null <- data.frame(fread("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/height_2018_30124842_hg19_genePvaluesNullGwas.txt.gz", data.table = F), row.names = 1)

gwas.x.null.norm <- data.frame(fread("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/genePvaluesNullGwas_as_zscores.txt.gz", data.table = F), row.names = 1)

# Transformed gwas zscores (centered and scaled) for aggregated genes
gwas <- data.frame(fread("~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/Coregulation_Enrichment_normalizedGwasGeneScores_ExHla.txt", data.table = F), row.names = 1)

# GWAS null pvalues
gwas.null<- data.frame(fread("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/Coregulation_Enrichment_normalizedNullGwasGeneScores_ExHla.txt.gz", data.table = F), row.names = 1)

# Cene coregulation zscores
coreg <- data.frame(fread("~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/Coregulation_Enrichment_normalizedPathwayScores_ExHla.txt.gz", data.table = F), row.names = 1)
```

# LOADING: Chr1q
```{r}
# GWAS pvalues als Z-score die gecentered and gescaled is
gwas.chr1q <- data.frame(fread("~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_1_q_Enrichment_geneScores.txt.gz", data.table = F), row.names = 1)

# Cene coregulation zscores
coreg.chr1q <- data.frame(fread("~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_1_q_Enrichment_pathwayScores.txt.gz", data.table = F), row.names = 1)

# Cene gene correlations
gene.cor.chr1q <- data.frame(fread("~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_1_q_Enrichment_geneCor.txt.gz", data.table = F), row.names = 1)
```

# LOADING: Gene gene correlation matrix
```{r}
gene.cor <- matrix(rep(0, 7713^2), ncol=7713, nrow=7713)
gene.names <- c()

index <- 1
for (chr in 1:22) {
  tmp.mat.q <- data.frame(fread(paste0("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_", chr, "_q_Enrichment_geneCor.txt.gz"), data.table = F), row.names = 1)

  tmp.mat.p <- tryCatch(
        {
          data.frame(fread(paste0("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_", chr, "_p_Enrichment_geneCor.txt.gz"), data.table = F), row.names = 1)
        },
        error=function(cond) {
            return(NA)
        })

  if (class(tmp.mat.p) == "data.frame") {
      tmp.range <- index:((index + nrow(tmp.mat.p))-1)
      gene.cor[tmp.range, tmp.range] <- as.matrix(tmp.mat.p)
      index <- index + nrow(tmp.mat.p)
      gene.names <- c(gene.names, rownames(tmp.mat.p))
  }

  tmp.range <- index:((index + nrow(tmp.mat.q))-1)
  gene.cor[tmp.range, tmp.range] <- as.matrix(tmp.mat.q)

  index <- index + nrow(tmp.mat.q)
  gene.names <- c(gene.names, rownames(tmp.mat.q))
}

rownames(gene.cor) <- gene.names
colnames(gene.cor) <- gene.names
```

# LOADING: Inverse Gene gene correlation matrix
```{r}
# Empty matrix of zeroes
gene.cor.inv <- matrix(rep(0, 7713^2), ncol=7713, nrow=7713)
gene.names <- c()

# Used to keep track of current row number, is inclusive 1 indexed
index <- 1
for (chr in 1:22) {
  
  # Read matrix for q arm
  tmp.mat.q <- data.frame(fread(paste0("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_", chr, "_q_Enrichment_geneInvCor.txt.gz"), data.table = F), row.names = 1)
  
  # Make sure order is identical
  tmp.mat.q <- tmp.mat.q[rownames(tmp.mat.q), rownames(tmp.mat.q)]

  # Define the range of the matrix to fill
  # -1 because R indexing is inclusive and 1 indexed
  tmp.range <- index:((index + nrow(tmp.mat.q))-1)
  gene.cor.inv[tmp.range, tmp.range] <- as.matrix(tmp.mat.q)
  
  # Advance the index so the next matrix will be 1 diagonal down
  index <- index + nrow(tmp.mat.q)
  gene.names <- c(gene.names, rownames(tmp.mat.q))
  
  # Attempt to read matrix for P arm
  tmp.mat.p <- tryCatch({
          data.frame(fread(paste0("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_", chr, "_p_Enrichment_geneInvCor.txt.gz"), data.table = F), row.names = 1)
        },
        error=function(cond) {
            return(NA)
        })
  
  if (class(tmp.mat.p) == "data.frame") {
      # Same as above
      tmp.mat.p <- tmp.mat.p[rownames(tmp.mat.p), rownames(tmp.mat.p)]
    
      tmp.range <-index:((index + nrow(tmp.mat.p))-1)
      gene.cor.inv[tmp.range, tmp.range] <- as.matrix(tmp.mat.p)
      index <- index + nrow(tmp.mat.p)
      gene.names <- c(gene.names, rownames(tmp.mat.p))
  }
}

rownames(gene.cor.inv) <- gene.names
colnames(gene.cor.inv) <- gene.names
```

# Comparing DEPICT inversion and R inversion
```{r}
tmp <- solve(gene.cor)

png("~/Desktop/depict2/plots/pairs_plot_inverse_cor_matrix_depict_r_comparison.png", width=2000, height=2000)
par(mfrow=c(10,10))
for (i in 15000:16000) {
  plot(tmp[,i], gene.cor.inv[,i], main=colnames(gene.cor.inv)[i], xlab=NA, ylab=NA)
}
dev.off()

# Most values are identical, some not probably due to double precision stuff
sum(tmp==gene.cor.inv)
```

# Comparing distributions of GWAS gene zscores
```{r}
par(mfrow=c(1,3))
ol <- intersect(rownames(gwas.x), rownames(gwas.chr1q))
plot(-log10(gwas.x[ol, 1]), gwas.chr1q[ol,1], main = "-log10 vs centered and scaled Zscore")
hist(gwas.chr1q[,1], breaks=100, xlab="Centered and scaled GWAS gene Zscores", main="Centered and scaled zscore")
hist(-log10(gwas.x[ol,1]), breaks=100, xlab="-log10 GWAS P", main="-log10 gene p")

ol <- intersect(rownames(gwas.null), rownames(gwas))

par(mfrow=c(1,3))
hist(gwas[,1], breaks=100, xlab="Centered and scaled GWAS gene Zscores", main="Actual")
hist(gwas.null[ol,1], breaks=100, xlab="Centered and scaled GWAS gene Zscores", main="Random 1")
hist(gwas.null[ol,10000], breaks=100, xlab="Centered and scaled GWAS gene Zscores", main="Random 10.000")

par(mfrow=c(1,4))
hist(gwas.x.null.norm[,1], breaks=100, xlab="GWAS gene Zscores", main="Random 1")
hist(scale(gwas.x.null.norm[,1]), breaks=100, xlab="Centerd and scaled GWAS gene Zscores", main="Random 1")
hist(gwas.x.null.norm[,10000], breaks=100, xlab="GWAS gene Zscores", main="Random 10.000")
hist(scale(gwas.x.null.norm[,10000]), breaks=100, xlab="Centerd and scaled GWAS gene Zscores", main="Random 10000")


```

# OLS
```{r}
par(mfrow=c(1,1))
plot(gwas.chr1q[,1], coreg.chr1q[,"ENSG00000075420"])

m.ols <- lm(coreg[,"ENSG00000075420"] ~ gwas[,1])
plot(gwas[,1], coreg[,"ENSG00000075420"])
abline(m.ols, col="blue")
```

# GLS
```{r}
# Chr1q
x     <- cbind(rep(1, nrow(gwas.chr1q)), coreg.chr1q[,1])
y     <- gwas.chr1q[,1]
Sigma <- gene.cor.chr1q

# Test gene ENSG00000075420
x     <- coreg[,"ENSG00000000003"]
y     <- gwas[rownames(coreg),1]
#Sigi  <- solve(gene.cor)
Sigi  <- gene.cor.inv[rownames(coreg), rownames(coreg)]
gls.out <- custom.gls(y, x, Sigi)

# Cleanup to save mem
rm(Sigma, gene.cor, gene.cor.chr1q, gwas.chr1q, tmp.mat.q, x, gwas.null, gwas.x, tmp.mat.p, tmp.range, index, gene.names)

gls.out <- sapply(colnames(coreg)[1:1000], function(gene){
  return(custom.gls(gwas[,1], coreg[,gene], Sigi))
})
gls.out <- t(gls.out)
rownames(gls.out) <- colnames(coreg)[1:1000]

ols.out <- sapply(colnames(coreg)[1:1000], function(gene){
  return(summary(lm(coreg[,gene] ~ gwas[,1]))$coefficients[2,])
})
ols.out <- t(ols.out)
rownames(ols.out) <- colnames(coreg)[1:1000]

ol <- intersect(rownames(gls.out), rownames(betas))

par(mfrow=c(1,3))
plot(betas[ol,1], gls.out[ol,1], xlab="Betas DEPICT2", ylab="R GLS model", main=paste0("Beta comparrsion ", length(ol), " genes"), xlim=c(-0.1, 0.1), ylim=c(-0.1, 0.1))
abline(a=0, b=1, col="red")
abline(lm(gls.out[ol,1] ~ betas[ol,1]), col="blue")

plot(betas[ol,1], ols.out[ol,1], xlab="Betas DEPICT2", ylab="R OLS model", main=paste0("Beta comparrsion ", length(ol), " genes"), xlim=c(-0.1, 0.1), ylim=c(-0.1, 0.1))
abline(a=0, b=1, col="red")
abline(lm(ols.out[ol,1] ~ betas[ol,1]), col="blue")

plot(gls.out[ol,1], ols.out[ol,1], xlab="R GLS model", ylab="R OLS model", main=paste0("Beta comparrsion ", length(ol), " genes"), xlim=c(-0.1, 0.1), ylim=c(-0.1, 0.1))
abline(a=0, b=1, col="red")
abline(lm(ols.out[ol,1] ~ gls.out[ol,1]), col="blue")

par(mfrow=c(1,1))

boxplot(betas[ol,1], gls.out[ol,1], ols.out[ol,1], names = c("DEPICT", "GLS R", "OLS R"))
abline(h=0, lty=2)

df.plot <- data.frame(c(betas[ol,1], gls.out[ol,1], ols.out[ol,1]))
df.plot$source <- c(rep("DEPICT", 1000), rep("GLS R", 1000), rep("OLS R", 1000))
colnames(df.plot) <- c("beta", "source")

df.plot<- df.plot[df.plot$source != "DEPICT", ]

p <- ggplot(aes(x=beta,fill=source, color=source), data=df.plot) +
  geom_histogram(position="identity",  alpha=0.5, bins=50, color="lightgrey") 

theme.nature(p) + geom_vline(xintercept = 0, lty=2,)+ scale_fill_manual(values=c("red", "blue"))

library(gridExtra)
grid.arrange(grobs=plots, ncol=3)

```

# Beta hist
```{r}
par(mfrow=c(1,4))
betas <- data.frame(fread("~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/Coregulation_Enrichment_betasExHla.txt",data.table=F), row.names=1)
betas.null <- data.frame(fread("~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/Coregulation_EnrichmentNull_betasExHla.txt",data.table=F), row.names=1)


df.plot <- data.frame(c(betas[,1], unlist(betas.null[sample.int(n=10000, size=1000)])))
df.plot <- cbind(df.plot, c(rep("Actual", nrow(betas)), rep("1000 random", nrow(betas)*1000)))
colnames(df.plot) <- c("beta", "source")

p <- ggplot(aes(x=beta, fill=source), data=df.plot) + geom_histogram(aes(y = ..density..), position="identity", alpha=0.5, bins=100)
theme.nature(p)

```

# Calculate beta using depict model; calculate beta's per chromosome arm
```{r}
b1.all <- c()
b2.all <- c()
gene.names <- c()

b1.all.depict <- c()
b2.all.depict <- c()

for (chr in 1:22) {
  for(arm in c("p", "q")) {
    
      Sigi <- tryCatch({
          data.frame(fread(paste0("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_", chr, "_", arm ,"_Enrichment_geneInvCor.txt.gz"), data.table = F), row.names = 1)
      }, error=function(cond) {
          return(NA)
      })

      x <- tryCatch({
        data.frame(fread(paste0("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_", chr, "_", arm ,"_Enrichment_geneScores.txt.gz"), data.table = F), row.names = 1)
      }, error=function(cond) {
          return(NA)
      })
      
      y <- tryCatch( {
        data.frame(fread(paste0("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_", chr, "_", arm ,"_Enrichment_pathwayScores.txt.gz"), data.table = F), row.names = 1)
      }, error=function(cond) {
          return(NA)
      })
      
      b1.depict <- tryCatch( {
        data.frame(fread(paste0("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_", chr, "_", arm ,"_Enrichment_b1.txt.gz"), data.table = F), row.names = 1)
      }, error=function(cond) {
          return(NA)
      })
      
      b2.depict <- tryCatch( {
        data.frame(fread(paste0("zcat ~/Desktop/depict2/output/height_paper/debug_files/height_2018_30124842_hg19_51/per_arm/Coregulation_", chr, "_", arm ,"_Enrichment_b2.txt.gz"), data.table = F), row.names = 1)
      }, error=function(cond) {
          return(NA)
      })

      
  if (class(Sigi) == "data.frame") {
      Sigi <- Sigi[rownames(y), rownames(y)]
      x   <- x[rownames(y),]
    
      out        <- custom.gls.per.arm(x, y[,1], as.matrix(Sigi))
      b1.all     <- c(b1.all, out[1])
      b2.all     <- c(b2.all, out[2])
      gene.names <- c(gene.names, colnames(y)[1])
      
      b1.all.depict <- c(b1.all.depict, b1.depict[1,2])
      b2.all.depict <- c(b2.all.depict, b2.depict[1,colnames(y)[1]])

  }
  }
}

plot(b1.all, b1.all.depict)
plot(b2.all, b2.all.depict)

sum(b2.all) / sum(b1.all) 


```

